src/main/java/com/google/sps/DataServlet.java

```diff
@@ -55,73 +119,111 @@ public void doGet(HttpServletRequest request, HttpServletResponse response) thro
+    // Truncated version of graph to return to the client
+    MutableGraph<GraphNode> truncatedGraph;
 
-    // Initialize variables if any are null. Ideally should all be null or none
-    // should be null
-    if (currDataGraph == null && originalDataGraph == null) {
+    // The list of mutations that need to be applied to the nodes in currDataGraph
+    // to get the requested graph, filtered to only contain changes pertaining
+    // to nodes in the truncated graph or the filtered node (null if the graph
+    // requested is before the current graph in the sequence of mutations)
+    MultiMutation diff = null;
 
-      /*
-       * The below code is used to read a graph specified in textproto form
-       */
-      InputStreamReader graphReader =
-          new InputStreamReader(
-              getServletContext().getResourceAsStream("/WEB-INF/graph.textproto"));
-      Graph.Builder graphBuilder = Graph.newBuilder();
-      TextFormat.merge(graphReader, graphBuilder);
-      Graph protoGraph = graphBuilder.build();
-
-      Map<String, Node> protoNodesMap = protoGraph.getNodesMapMap();
-      // Originally both set to same data
-      originalDataGraph = DataGraph.create();
-      success = originalDataGraph.graphFromProtoNodes(protoNodesMap);
-      String error = "Failed to parse input graph into Guava graph - not a DAG!";
-      if (!success) {
-        response.setHeader("serverError", error);
-        return;
-      }
-      currDataGraph = originalDataGraph.getCopy();
-    } else if (currDataGraph == null || originalDataGraph == null) {
-      String error = "Invalid input";
-      response.setHeader("serverError", error);
-      return;
-    }
+    response.setContentType("application/json");
+    String graphJson;
 
-    // Mutations file hasn't been read yet
-    if (mutList == null) {
-      /*
-       * The below code is used to read a mutation list specified in textproto form
-       */
-      InputStreamReader mutReader =
-          new InputStreamReader(
-              getServletContext().getResourceAsStream("/WEB-INF/mutation.textproto"));
-      MutationList.Builder mutBuilder = MutationList.newBuilder();
-      TextFormat.merge(mutReader, mutBuilder);
-      mutList = mutBuilder.build().getMutationList();
+    /*
+     *************************************************
+     * Filtering Graph and Mutations By Searched Node
+     *************************************************
+     */
+
+    // Get the diff if we are going forward in the list of mutations
+    if (mutationNumber > currDataGraph.numMutations()) {
+      diff = Utility.getDiffBetween(mutList, mutationNumber);
     }
+    // Find the indices that mutate the searched node, computing and caching them
+    // if this has not been done already
+    if (!mutationIndicesMap.containsKey(nodeNameParam)) {
+      mutationIndicesMap.put(
+          nodeNameParam, Utility.getMutationIndicesOfNode(nodeNameParam, mutList));
```

Comment:
the alternative would be to not store it. then if they search the same node that doesn't exist again, it would have to get the indices again which seems like not great either. 

What do you think?